scipy/interpolate/src/__fitpack.cc

```diff
@@ -0,0 +1,426 @@
+    https://github.com/scipy/scipy/blob/maintenance/1.11.x/scipy/interpolate/fitpack/fpcurf.f#L112-L161
+    A possibly more efficient way could be to note that all data points which
+    lie between two knots all have the same offset: if `t[i] < x_1 .... x_s < t[i+1]`,
+    the `s-1` corresponding rows form an `(s-1, k+1)`-sized "block".
+    Then a blocked QR implementation could look like
+    https://people.sc.fsu.edu/~jburkardt/f77_src/band_qr/band_qr.f
+
+    The `startrow` optional argument accounts for the scenatio with a two-step
+    factorization. Namely, the preceding rows are assumend to be already
+    processed and are skipped.
+    This is to account for the scenario where we append new rows to an already
+    triangularized matrix.
+
+    This routine MODIFIES `a` & `y` in-place.
+
+ */
+void
+qr_reduce(double *aptr, const ssize_t m, const ssize_t nz, // a(m, nz), packed
+          ssize_t *offset,                                 // offset(m)
+          const ssize_t nc,                                // dense would be a(m, nc)
+          double *yptr, const ssize_t ydim1,               // y(m, ydim2)
+          const ssize_t startrow
+)
+{
+    auto R = RealArray2D(aptr, m, nz);
+    auto y = RealArray2D(yptr, m, ydim1);
+
+    for (ssize_t i=startrow; i < m; ++i) {
+        ssize_t oi = offset[i];
+        for (ssize_t j=oi; j < nc; ++j) {
+
+            // rotate only the lower diagonal
+            if (j >= std::min(i, nc)) {
+                break;
+            }
+
+            // in dense format: diag a1[j, j] vs a1[i, j]
+            double c, s, r;
+            DLARTG(&R(j, 0), &R(i, 0), &c, &s, &r);
+
+            // rotate l.h.s.
+            R(j, 0) = r;
+            for (ssize_t l=1; l < R.ncols; ++l) {
+                std::tie(R(j, l), R(i, l-1)) = fprota(c, s, R(j, l), R(i, l));
+            }
+            R(i, R.ncols-1) = 0.0;
+
+            // rotate r.h.s.
+            for (ssize_t l=0; l < y.ncols; ++l) {
+                std::tie(y(j, l), y(i, l)) = fprota(c, s, y(j, l), y(i, l));
+            }
+        }
+        if (i < nc) {
+            offset[i] = i;
+        }
+
+    } // for(i = ...
+}
+
+
+/*
+ * Back substitution solve of `R @ c = y` with an upper triangular R.
+ *
+ * R is in the 'packed' format:
+ *    1. Each row has at most `nz` non-zero elements, stored in a (m, nz) array
+ *    2. All non-zeros are consecutive.
+ * Since R is upper triangular, non-zeros in row `i` start at `i`. IOW,
+ * the 'offset' array is `np.arange(nc)`.
+ *
+ * If `R` were dense, it would have had shape `(m, nc)`. Since R is upper triangular,
+ * the first `nc` rows contain non-zeros; the last `m-nc` rows are all zeros
+ * (in fact, they may contain whatever, and are not referenced in the routine).
+ *
+ * `y` array is always two-dimensional, and has shape `(m, ydim2)`.
+ * IOW, if the original data is 1D, `y.shape == (m, 1)`.
+ *
+ * The output `c` array has the shape `(nc, ydim2)`.
+ */
+void
+fpback( /* inputs*/
```

Comment:
Maybe I read a bit too quickly but isn't [`dtpsv`](https://www.netlib.org/lapack/explore-html/d7/d3b/group__tpsv_gaa7aa06a5635bdd236463a56a889d3aa4.html#gaa7aa06a5635bdd236463a56a889d3aa4) does this since you are using LAPACK?

Reply:
Quite possibly, thanks! My "packed" format is not exactly the LAPACK packed format, so it'll require a bit of work to get right.

Reply:
Actually, no. The "packed" format here only keeps no more than nz=k+2 elements per row, so that the R array here has max size (m*nz) not m(m+1) /2.  (with m >> k). So I can in principle copy, call LAPACK and copy back but then why paying both copying and big-O.

Reply:
Ah OK, so when you mention packed storage, it's not LAPACK's packed storage. Got it.